충남대 BE 김재혁 1단계

dmdkr55 opened the pull request:
안녕하세요 멘토님.
위시리스트 1단계 - 유효성 검사 및 예외 처리 과제를 완료하여 코드 리뷰 요청드립니다.
감사합니다!

dmdkr55 committed:
feat: 기본 코드 준비

dmdkr55 committed:
feat: 유효성 검사

dmdkr55 committed:
feat: 예외 처리

- GlobalExceptionHandler
예외 발생시 응답 메시지 제공

dmdkr55 committed:
feat: 테스트 코드 추가

- ProductControllerTest
유효성 검사 및 예외 처리에 대한 테스트 코드 추가

dmdkr55 committed:
Merge branch 'dmdkr55' into step1

sumniy commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
오 `ResponseEntityExceptionHandler`를 상속하게 구현하셨네요!  어떤 기능들을 제공하나요?

sumniy commented on src/main/java/gift/model/Product.java:
```diff
@@ -1,23 +1,38 @@
+    @PositiveOrZero(message = "가격은 음수가 될 수 없습니다.")
+    private Integer price;
+
+    @NotBlank(message = "이미지 Url은 필수 입력 값입니다.")
```
`@Pattern` 어노테이션을 활용해서 url 패턴만 허용하도록 할 수 있어요!

sumniy commented on src/main/java/gift/validation/ProductNameValidator.java:
```diff
@@ -0,0 +1,17 @@
+public class ProductNameValidator implements ConstraintValidator<ValidProductName, String> {
+
+    private static final String ALLOWED_SPECIAL_CHARS = "()+\\-\\[\\]&/_";
+    private static final String PATTERN = "^[a-zA-Z0-9" + ALLOWED_SPECIAL_CHARS + "가-힣\\s]*$";
```
상수로 의미 구분해서 가독성이 좋아졌네요 👍
참고로 저는 이렇게 바로 쉽게 읽히지 않는 부분들은 주석을 활용해서 가독성을 조금이라도 높이는 걸 선호합니다 ㅎㅎ
```suggestion
    private static final String ALLOWED_SPECIAL_CHARS = "()+\\-\\[\\]&/_"; // ( ), [ ], +, -, &, /, _
    private static final String PATTERN = "^[a-zA-Z0-9" + ALLOWED_SPECIAL_CHARS + "가-힣\\s]*$";
```

sumniy commented on src/main/java/gift/validation/ProductNameValidator.java:
```diff
@@ -0,0 +1,17 @@
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        if (value == null) return false;
+        if (value.contains("카카오")) return false;
```
MD와 협의해서 '카카오 초콜릿' 와 같은 단어를 허용하고 싶다면 어떻게 해야할까요? 이런 단어들 목록이 자주 변경되거나 많아진다면, 현재 구조에서 수정하기 편할까요?

고민해 보시면 좋을 것 같아요 😊

sumniy commented on src/main/java/gift/validation/ValidProductName.java:
```diff
@@ -0,0 +1,20 @@
+@Constraint(validatedBy = ProductNameValidator.class)
+public @interface ValidProductName {
+
+    String message() default "특수 문자는 ((, ), [, ], +, -, &, /, _)만 가능합니다. "
```
명확한 메시지 좋네요 👍

sumniy commented on src/main/java/gift/validation/ValidProductName.java:
```diff
@@ -0,0 +1,20 @@
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = ProductNameValidator.class)
+public @interface ValidProductName {
```
커스텀 어노테이션 활용 💯

sumniy commented on src/test/java/gift/controller/ProductControllerTest.java:
```diff
@@ -55,6 +55,94 @@ void setUp() {
+        mockMvc.perform(post("/api/products").contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(product))).andExpect(status().isBadRequest())
+            .andExpect(jsonPath("$.messages[0]").value("이름은 필수 입력 값입니다.")); // 예외 메시지 일치 여부
+    }
```
테스트가 실패하네요~ 
디버깅해서 수정 부탁드립니다!

sumniy reviewed (changes requested):
안녕하세요 재혁님!

1단계 잘 구현해 주셨지만, 어드민 화면으로 테스트 해보면 상품명 유효성 검사가 작동하지 않네요!
수정 부탁드릴게요 ㅎㅎ 빠트리신 게 있을 거예요~

dmdkr55 committed:
feat: Admin 화면에서의 유효성 검사 추가

- ProductAdminController
유효성 검사 및 예외 처리에 대한 코드 추가
-create-form.html, edit-form.html
예외 처리에 따른 Exception 메시지

dmdkr55 committed:
Merge branch 'step1' of https://github.com/dmdkr55/spring-gift-wishlist into step1

dmdkr55 committed:
refactor: 이미지 URL 유효성 검사

- Product.java
이미지 URL에 대한 pattern을 지정

dmdkr55 committed:
refactor: 상품이름 유효성 검사 수정

- Product.java
'카카오'가 포함된 이름은 needsMdApproval 필드를 true로 하여, MD의 승인이 필요하도록 수정.

dmdkr55 committed:
refactor: 변경된 내용에 따라 테스트 코드 수정
          없는 id의 상품 단건 조회시 오류 수정.

- ProductControllerTest.java
수정된 유효성 검사에 따라 테스트 코드 수정
존재하지 않는 상품 조회에 대한 테스트 추가
- ProductRepository.java
존재하지 않는 상품 조회에 대한 에러 처리
